### PR TITLE
(WIP) Feature/ethereum transaction

### PR DESCRIPTION
**Pull request by Totoual**
## Proposed changes

Intergrade ethereum ledger. Example using ethereum.

Addresses #329
## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../master/CONTRIBUTING.rst) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/crypto/helpers.py [hunk 1]
```diff
@@ -145,8 +145,8 @@ def _verify_ledger_apis_access(ctx: Context) -> None:
     else:
         ethereum_ledger_config = cast(LedgerAPIConfig, ethereum_ledger_config)
         try:
-            raise NotImplementedError
-            # TODO connect to ledger
+            from web3 import Web3, HTTPProvider     # type: ignore
```
Comment (DavidMinarsch) [inline aea/crypto/helpers.py]: can we add web3 to mypi ignore so it is globally ignored?

File: aea/crypto/ledger_apis.py [hunk 2]
```diff
@@ -25,12 +25,17 @@
 
 from fetchai.ledger.api import LedgerApi as FetchLedgerApi  # type: ignore
 # from fetchai.ledger.api.tx import TxStatus
-from fetchai.ledger.crypto import Entity, Identity, Address  # type: ignore
+from fetchai.ledger.crypto import Identity, Address  # type: ignore
```
Comment (DavidMinarsch) [inline aea/crypto/ledger_apis.py]: let's add fetchai also to mypi ignore. Then the type: ignore can be removed throughout.

File: aea/crypto/ledger_apis.py [hunk 3]
```diff
@@ -25,12 +25,17 @@
...
+from web3 import Web3, HTTPProvider  # type: ignore
 
+from aea.crypto.base import Crypto
 from aea.crypto.ethereum import ETHEREUM
 from aea.crypto.fetchai import FETCHAI
 
 DEFAULT_FETCHAI_CONFIG = ('alpha.fetch-ai.com', 80)
+ETHEREUM_TEST_NETWORK = ("https://ropsten.infura.io/v3/f00f7b3ba0e848ddbdc8941c527447fe",)
```
Comment (DavidMinarsch) [inline aea/crypto/ledger_apis.py]: do we really want to commit this to the AEA package? Put differently, is this expected to work for the foreseeable future?

File: aea/crypto/ledger_apis.py [hunk 4]
```diff
@@ -69,33 +75,62 @@ def token_balance(self, identifier: str, address: str) -> int:
...
+                    'nonce': 0x05,
+                    'chainId': 3,
+                    'to': api.toChecksumAddress(destination_address),
+                    'value': amount,
+                    'gas': tx_fee + api.eth.gasPrice,
+                    'gasPrice': api.eth.gasPrice
+                }
+                signature = api.eth.account.signTransaction(transaction_dict=transaction,
+                                                            private_key=crypto_object.entity.key)
+                tx_digest = api.eth.sendRawTransaction(signature.rawTransaction)
```
Comment (DavidMinarsch) [inline aea/crypto/ledger_apis.py]: can we add similar log messages to the fetchai case please.

File: aea/crypto/ledger_apis.py [hunk 5]
```diff
@@ -110,25 +145,30 @@ def is_tx_settled(self, identifier: str, tx_digest: str, amount: int) -> bool:
...
+                # tx_status = cast(TxStatus, api.tx.status(tx_digest))
+                tx_status = cast(str, api.tx.status(tx_digest))
+                if tx_status in SUCCESSFUL_TERMINAL_STATES:
+                    # TODO: check the amount of the transaction is correct
+                    is_successful = True
+                logger.info("Transaction validated ...")
+            except Exception:
+                logger.warning("An error occurred while attempting to check the transaction.")
+        elif identifier == ETHEREUM:
+            is_successful = True
```
Comment (DavidMinarsch) [inline aea/crypto/ledger_apis.py]: Can you add the code for ethereum here?

File: aea/crypto/ledger_apis.py [hunk 6]
```diff
@@ -110,25 +145,30 @@ def is_tx_settled(self, identifier: str, tx_digest: str, amount: int) -> bool:
...
+
         return is_successful
 
     @staticmethod
     def get_address_from_public_key(self, identifier: str, public_key: str) -> Address:
         """
         Get the address from the public key.
 
+        :param self:
```
Comment (DavidMinarsch) [inline aea/crypto/ledger_apis.py]: we don't add self in docstrings :)

File: aea/crypto/ledger_apis.py [hunk 7]
```diff
@@ -69,33 +75,66 @@ def token_balance(self, identifier: str, address: str) -> int:
...
+                    'to': api.toChecksumAddress(destination_address),
+                    'value': amount,
+                    'gas': tx_fee + api.eth.gasPrice,
+                    'gasPrice': api.eth.gasPrice
+                }
+
+                signature = api.eth.account.sign_transaction(transaction_dict=transaction,
+                                                             private_key=crypto_object.entity.privateKey)
+
+                logger.info("Signature: {}".format(signature))
```
Comment (DavidMinarsch) [inline aea/crypto/ledger_apis.py]: Info should not print this. The user does not need to know this. You could print as part of `debug`. Please add the same `info` messages as above in fetchai case.

File: aea/crypto/ledger_apis.py (hunk 5 above)
Comment (Totoual) [reply aea/crypto/ledger_apis.py]: I will once I solve the problem with the transaction.

File: packages/skills/weather_client_ledger/handlers.py [hunk 8]
```diff
@@ -206,20 +210,22 @@ def _handle_match_accept(self, msg: FIPAMessage, sender: str, message_id: int, d
         logger.info("[{}]: received MATCH_ACCEPT_W_ADDRESS from sender={}".format(self.context.agent_name, sender[-5:]))
         address = cast(str, msg.get("address"))
         proposal = cast(Description, dialogue.proposal)
+        crypto = CURRENCY_TO_ID_MAP[proposal.values.get("currency_pbk")]
```
Comment (DavidMinarsch) [inline packages/skills/weather_client_ledger/handlers.py]: Same as above

File: packages/skills/weather_client_ledger/handlers.py [hunk 9]
```diff
@@ -206,20 +210,22 @@ def _handle_match_accept(self, msg: FIPAMessage, sender: str, message_id: int, d
...
         tx_msg = TransactionMessage(performative=TransactionMessage.Performative.PROPOSE,
                                     skill_id="weather_client_ledger",
                                     transaction_id="transaction0",
-                                    sender=self.context.agent_public_keys['fetchai'],
+                                    sender=self.context.agent_public_keys[crypto],
                                     counterparty=address,
                                     is_sender_buyer=True,
-                                    currency_pbk="FET",
+                                    currency_pbk=crypto,
```
Comment (DavidMinarsch) [inline packages/skills/weather_client_ledger/handlers.py]: the currency_pbk should be the `currency_pbk` not the `ledger_id`

File: packages/skills/weather_station_ledger/strategy.py [hunk 10]
```diff
@@ -86,7 +86,7 @@ def generate_proposal_and_data(self, query: Query) -> Tuple[Description, Dict[st
         total_price = self.price_per_row * rows
         proposal = Description({"rows": rows,
                                 "price": total_price,
-                                "currency_pbk": self.currency})
+                                "currency_pbk": self.currency[1]})
```
Comment (DavidMinarsch) [inline packages/skills/weather_station_ledger/strategy.py]: In the configs for strategy specify the `ledger_id` and then load it in the constructor and choose accordingly.

File: tests/test_decision_maker/test_base.py [hunk 11]
```diff
@@ -64,20 +64,20 @@ def test_initialisation(self):
 
     def test_transaction_is_consistent(self):
         """Test the consistency of the transaction message."""
-        currency_endowment = {"FET": 100.0}
-        good_endowment = {"FET": 2}
+        currency_endowment = {"fetchai": 100.0}
```
Comment (DavidMinarsch) [inline tests/test_decision_maker/test_base.py]: Please revert all to `currency_pbk`.

File: packages/skills/weather_client_ledger/handlers.py (hunk 9 above)
Comment (Totoual) [reply packages/skills/weather_client_ledger/handlers.py]: This is based on the logic you had on the decision_maker. You were choosing which address to use based on the currency_pbk.

File: packages/skills/weather_station_ledger/strategy.py [hunk 12]
```diff
@@ -51,6 +52,7 @@ def __init__(self, **kwargs) -> None:
         """
         self.price_per_row = kwargs.pop('price_per_row') if 'price_per_row' in kwargs.keys() else DEFAULT_PRICE_PER_ROW
         self.currency = kwargs.pop('currency_pbk') if 'currency_pbk' in kwargs.keys() else DEFAULT_CURRENCY_PBK
```
Comment (DavidMinarsch) [inline packages/skills/weather_station_ledger/strategy.py]: please call this `self.currency_pbk`

File: packages/skills/weather_station_ledger/strategy.py [hunk 13]
```diff
@@ -86,7 +88,8 @@ def generate_proposal_and_data(self, query: Query) -> Tuple[Description, Dict[st
         total_price = self.price_per_row * rows
         proposal = Description({"rows": rows,
                                 "price": total_price,
-                                "currency_pbk": self.currency})
+                                "currency_pbk": self.currency[1],
```
Comment (DavidMinarsch) [inline packages/skills/weather_station_ledger/strategy.py]: self.currency_pbk should hold a string, not a list

File: packages/skills/weather_station_ledger/skill.yaml [hunk 14]
```diff
@@ -17,7 +17,8 @@ shared_classes:
       class_name: Strategy
       args:
         price_per_row: 1
-        currency_pbk: 'FET'
+        currency_pbk: ['FET', 'ETH']
```
Comment (DavidMinarsch) [inline packages/skills/weather_station_ledger/skill.yaml]: this should be either `FET` or `ETH`, in this case `ETH` as `ledger_id` is `ethereum`

File: packages/skills/weather_station_ledger/handlers.py [hunk 15]
```diff
@@ -205,7 +207,7 @@ def _handle_accept(self, msg: FIPAMessage, sender: str, message_id: int, dialogu
                                        dialogue_id=dialogue_id,
                                        target=new_target,
                                        performative=FIPAMessage.Performative.MATCH_ACCEPT_W_ADDRESS,
-                                       address=self.context.agent_addresses['fetchai'])
+                                       address=self.context.agent_addresses[USED_ADDRESS])
```
Comment (DavidMinarsch) [inline packages/skills/weather_station_ledger/handlers.py]: the dialogue needs to keep track of the `currency_pbk` and the `ledger_id`

File: packages/skills/weather_client_ledger/handlers.py [hunk 16]
```diff
@@ -206,20 +209,24 @@ def _handle_match_accept(self, msg: FIPAMessage, sender: str, message_id: int, d
         logger.info("[{}]: received MATCH_ACCEPT_W_ADDRESS from sender={}".format(self.context.agent_name, sender[-5:]))
         address = cast(str, msg.get("address"))
         proposal = cast(Description, dialogue.proposal)
+        ledger_id = cast(str, proposal.values.get("ledger_id"))
+        print(ledger_id)
```
Comment (DavidMinarsch) [inline packages/skills/weather_client_ledger/handlers.py]: please remove the print statement

File: aea/decision_maker/base.py [hunk 17]
```diff
@@ -326,7 +326,11 @@ def _handle_tx_message(self, tx_message: TransactionMessage) -> None:
         :return: None
         """
         # get variables
-        crypto_identifier = CURRENCY_TO_ID_MAP.get(cast(str, tx_message.get("currency_pbk")))
+        if tx_message.get("ledger_id") is not None:
```
Comment (DavidMinarsch) [inline aea/decision_maker/base.py]: ok maybe get it first and then check if it is not none

File: aea/decision_maker/base.py [hunk 18]
```diff
@@ -326,7 +326,11 @@ def _handle_tx_message(self, tx_message: TransactionMessage) -> None:
...
+            crypto_identifier = tx_message.get("ledger_id")
+        else:
```
Comment (DavidMinarsch) [inline aea/decision_maker/base.py]: you don't need that if you have the `ledger_id`

File: aea/decision_maker/messages/transaction.py [hunk 19]
```diff
@@ -52,6 +52,7 @@ def __init__(self, performative: Union[str, Performative],
                  quantities_by_good_pbk: Dict[str, int],
                  transaction_digest: Optional[str] = None,
                  dialogue_label: Optional[Dict[str, str]] = None,
+                 ledger_id: Optional[str] = None,
```
Comment (DavidMinarsch) [inline aea/decision_maker/messages/transaction.py]: There also needs to be a check in the consistency checks

**Comment by DavidMinarsch** [comment]
LGTM

**Pull request merged**